fusequery/query/src/catalogs/impls/database_catalog.rs

```diff
@@ -98,106 +110,96 @@ impl DatabaseCatalog {
     }
 }
 
-impl DatabaseCatalog {
-    pub fn get_database(&self, db_name: &str) -> Result<Arc<dyn Database>> {
-        let db_lock = self.databases.read();
-        let database = db_lock.get(db_name).ok_or_else(|| {
-            ErrorCode::UnknownDatabase(format!("Unknown database: '{}'", db_name))
-        })?;
-        Ok(database.clone())
+#[async_trait::async_trait]
+impl Catalog for DatabaseCatalog {
+    fn get_database(&self, db_name: &str) -> Result<Arc<dyn Database>> {
+        self.databases.read().get(db_name).map_or_else(
+            || {
+                if !self.disable_remote {
+                    self.meta_store_cli.get_database(db_name)
+                } else {
+                    Err(ErrorCode::UnknownDatabase(format!(
+                        "Unknown database {}",
+                        &db_name
+                    )))
+                }
+            },
+            |v| Ok(v.clone()),
+        )
     }
 
-    pub fn get_databases(&self) -> Result<Vec<String>> {
-        let mut results = vec![];
-        for (k, _v) in self.databases.read().iter() {
-            results.push(k.clone());
+    fn get_databases(&self) -> Result<Vec<String>> {
+        let locals = self.databases.read();
+
+        if self.disable_remote {
+            return Ok(locals
+                .iter()
+                .map(|(k, _v)| k.to_owned())
+                .collect::<Vec<_>>());
         }
-        Ok(results)
-    }
 
-    pub fn get_table(&self, db_name: &str, table_name: &str) -> Result<Arc<TableMeta>> {
-        let db_lock = self.databases.read();
-        let database = db_lock.get(db_name).ok_or_else(|| {
-            ErrorCode::UnknownDatabase(format!("Unknown database: '{}'", db_name))
-        })?;
+        // merge with remote meta data
+        let locals = locals.iter().map(|(k, _v)| k).collect::<HashSet<_>>();
+        let remotes = self.meta_store_cli.get_databases()?;
+        let remotes = remotes.iter().collect::<HashSet<_>>();
+        let db_names = remotes.union(&locals);
+        let mut r = db_names.into_iter().cloned().collect::<Vec<_>>();
+        r.sort();
+        Ok(r.into_iter().cloned().collect())
+    }
```

Comment:
here in DatabaseCatalog, `remote databases` are not cached. only table_meta of specific id and version are cached.

Reply:
If dropping table is being concerned:

the cache of table_meta are NOT affected when tables are dropped, the cache is just a simple LRU cache.

When the plan builder access metadata of database/table, it will always(for the time being) request for  fresh ones, catalog will fetch them via RPC. only when a physical plan is being executed, a table of specific version is being asked for, at that time, table_meta cache might be hit.